code/modules/vehicles/mecha/equipment/mecha_equipment.dm

```diff
@@ -29,58 +31,43 @@
-/obj/item/mecha_parts/mecha_equipment/proc/update_chassis_page()
-	SHOULD_CALL_PARENT(TRUE)
-	send_byjax(chassis.occupants,"exosuit.browser","eq_list", chassis.get_equipment_list())
-	send_byjax(chassis.occupants,"exosuit.browser","equipment_menu", chassis.get_equipment_menu(),"dropdowns")
-	return TRUE
-
-///Updates chassis equipment list html menu with custom data
-/obj/item/mecha_parts/mecha_equipment/proc/update_equip_info()
-	if(!chassis)
-		return
-	send_byjax(chassis.occupants,"exosuit.browser","[REF(src)]",get_equip_info())
-	return TRUE
-
 /obj/item/mecha_parts/mecha_equipment/Destroy()
 	if(chassis)
-		LAZYREMOVE(chassis.equipment, src)
-		if(chassis.selected == src)
-			chassis.selected = null
-		update_chassis_page()
+		detach(get_turf(src))
 		log_message("[src] is destroyed.", LOG_MECHA)
 		if(LAZYLEN(chassis.occupants))
 			to_chat(chassis.occupants, "[icon2html(src, chassis.occupants)][span_danger("[src] is destroyed!")]")
 			playsound(chassis, destroy_sound, 50)
-		if(!detachable) //If we're a built-in nondetachable equipment, let's lock up the slot that we were in.
-			chassis.max_equip--
 		chassis = null
 	return ..()
 
-/obj/item/mecha_parts/mecha_equipment/try_attach_part(mob/user, obj/vehicle/sealed/mecha/mech)
-	if(!can_attach(mech))
-		to_chat(user, span_warning("You are unable to attach [src] to [mech]!"))
-		return FALSE
-	if(!user.temporarilyRemoveItemFromInventory(src))
-		return FALSE
-	attach(mech)
-	user.visible_message(span_notice("[user] attaches [src] to [mech]."), span_notice("You attach [src] to [mech]."))
-	return TRUE
-
-///fetches and returns a html formatted string with equippability status
-/obj/item/mecha_parts/mecha_equipment/proc/get_equip_info()
-	if(!chassis)
-		return
-	var/txt = "<span style=\"color:[activated?"#0f0":"#f00"];\">*</span>&nbsp;"
-	if(chassis.selected == src)
-		txt += "<b>[src]</b>"
-	else if(selectable)
-		txt += "<a href='?src=[REF(chassis)];select_equip=[REF(src)]'>[src]</a>"
-	else
-		txt += "[src]"
+/obj/item/mecha_parts/mecha_equipment/try_attach_part(mob/user, obj/vehicle/sealed/mecha/M, attach_right = FALSE)
+	if(can_attach(M, attach_right))
+		if(!user.temporarilyRemoveItemFromInventory(src))
+			return FALSE
+		attach(M, attach_right)
+		user.visible_message(span_notice("[user] attaches [src] to [M]."), span_notice("You attach [src] to [M]."))
+		return TRUE
+	to_chat(user, span_warning("You are unable to attach [src] to [M]!"))
+	return FALSE
 
-	return txt
+/obj/item/mecha_parts/mecha_equipment/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	switch(action)
+		if("detach")
+			detach(get_turf(src))
+			return TRUE
+		if("toggle")
+			activated = !activated
+			return TRUE
+		if("repair")
+			ui.close() // allow watching for baddies and the ingame effects
+			chassis.balloon_alert(usr, "starting repair")
+			while(do_after(usr, 1 SECONDS, chassis) && get_integrity() < max_integrity)
+				repair_damage(30)
```

Comment:
damn.. can't you do this in a different way like process()?
This seems like an easy way of doing it

Reply:
though that's a bit of an overexaggeration, it's not that bad but I feel like you could do this without a while() sleep() loop